### ISSUE TITLE
Possibility of process crash during startup leading to split-brain cluster and crash loop

### ISSUE THREAD

**Issue by brennanvincent** [MaterializeInc/materialize#12800]
### What version of Materialize are you using?

main

### How did you install Materialize?

Materialize Cloud

### What is the issue?

With the present release of timely, the following pattern can happen. Imagine there are three processes numbered 0, 1, and 2.

1. Process 0 comes up
2. Process 1 comes up and connects to process 0
3. Process 0 crashes for whatever reason (cosmic rays, Kubernetes gremlins, OOM, whatever)
4. Kubernetes brings up process 0 again.
5. Process 2 connects to process 0.
6. Process 2 connects to process 1 -- processes 1 and 2 have now completed initialization.
7. Process 1 still has a TCP handle to the old process 0, which is now closed; it therefore crashes.
8. Process 2 learns that process 1 has crashed, and itself crashes.

To recap, the state is now:

Process 0: Alive, believes it is connected to process 2
Process 1: Dead
Process 2: Dead

9. Process 1 and 2 come up, one of them (doesn't matter which) connects to process 0.
10. Process 0 now believes it has connected to two peers, but only one of the connections is valid, so it crashes.
11. We are now back in the state we were in after step 3 above. Thus, we can now crash-loop indefinitely.

**Comment by benesch** [comment]
I'm still trying to wrap my head around this. Is the problem that we don't check whether a prior connection has broken during initialization? If process 0 accepts a connection from process 1, but then process 1 crashes, process 0 won't attempt to read from the connection to process 1 _until_ it gets a connection from process 2, and therefore it won't notice that process 1 has crashed until it completes initialization?

**Comment by brennanvincent** [comment]
Right now that's correct, yes. We don't start reading until initialization is over. So one possible fix to this issue might be to crash earlier. (I'd still need to think about it in more detail to convince ourselves it converges to a good state)
